Add support for style=form explode=false to serialized parameter

ctipggroup opened the pull request:
Add support for style=form explode=false to serialized parameter

ctipggroup commented:
This is add support for

`
- in: query
          name: id
          style: form
          explode: false
          schema:
            type: array
            items:
              type: int
`

?id=1,2,3

scaytrase commented:
Can you please provide new test verifying that new validation works (both successful and failing cases)

ctipggroup commented:
> Can you please provide new test verifying that new validation works (both successful and failing cases)

Updated with tests. 
Thanks

ctipggroup commented on src/PSR7/Validators/SerializedParameter.php:
```diff
@@ -85,30 +94,47 @@ public function deserialize($value)
+    private function convertScalar($value, string $type)
+    {
+        if (($type === CebeType::BOOLEAN) && is_scalar($value) && preg_match('#^(true|false)$#i', (string) $value)) {
+            return strtolower($value) === 'true';
```
Fixed to return false for 'false'

ctipggroup commented:
> Can you please provide new test verifying that new validation works (both successful and failing cases)

Tests added
Added support to convert array items to correct item type
Fixed issue where 'false' is evaluated to true

ctipggroup commented on src/PSR7/Validators/SerializedParameter.php:
```diff
@@ -85,28 +93,51 @@ public function deserialize($value)
+     *
+     * @return mixed
+     */
+    private function convertScalar($value, ?string $type = '')
```
Removed

ctipggroup commented on src/PSR7/Validators/SerializedParameter.php:
```diff
@@ -85,28 +93,51 @@ public function deserialize($value)
 
-        if (($this->schema->type === CebeType::BOOLEAN) && is_scalar($value) && preg_match('#^(true|false)$#i', (string) $value)) {
-            return (bool) $value;
+        $value = $this->convertScalar($value, $this->schema->type);
```
Refactored as suggested.

ctipggroup commented on src/PSR7/Validators/SerializedParameter.php:
```diff
@@ -85,28 +93,51 @@ public function deserialize($value)
+            && is_string($value)) {
+            $value = explode(',', $value);
+            foreach ($value as &$val) {
+                $val = $this->convertScalar($val, $this->schema->items->type ?? '');
```
Changed to null as suggested

ctipggroup commented on src/PSR7/Validators/SerializedParameter.php:
```diff
@@ -85,28 +93,51 @@ public function deserialize($value)
+        $value = $this->convertScalar($value, $this->schema->type);
+
+        if (($this->schema->type === CebeType::ARRAY)
+            && ($this->style === 'form' && $this->explode === false)
```
Refactored as sugggested

ctipggroup commented on src/PSR7/Validators/SerializedParameter.php:
```diff
@@ -85,28 +93,51 @@ public function deserialize($value)
+        $value = $this->convertScalar($value, $this->schema->type);
+
+        if (($this->schema->type === CebeType::ARRAY)
+            && ($this->style === 'form' && $this->explode === false)
```
Constants added

scaytrase commented on tests/FromCommunity/IssueWithQueryArrayTest.php:
```diff
@@ -0,0 +1,112 @@
+    {
+        $map     = [
+            'form' => ['integer' => '1,2,3', 'string' => 'id1,id2,id3', 'boolean' => 'true,false', 'number' => '1.00,2.00,3.00'],
+            'spaceDelimited' => ['integer' => '1%202%203', 'string' => 'id1%20id2%20id3', 'boolean' => 'true%20false', 'number' => '1.00%202.00%203.00'],
```
I'm not sure whether real world requests have encoded values in request object

ctipggroup replied:
Updated

scaytrase commented:
Great work! Thanks @ctipggroup
